Add distributed process group manager

Delaunay opened the pull request.

breuleux commented on milatools/distributed/__init__.py:
```diff
@@ -0,0 +1,181 @@
+
+
+def record(fn, error_handler=None):
+    """Decorator that help record exception in a distrubted setup"""
```
```suggestion
    """Decorator that helps record an exception in a distributed setup"""
```

breuleux commented on milatools/distributed/__init__.py:
```diff
@@ -0,0 +1,181 @@
+
+def barrier():
+    """block until all workers reach this"""
+    return __get_attr("barrier", lambda: lambda: True)()
```
Did you mean the double lambda?

```suggestion
    return __get_attr("barrier", lambda: True)()
```

breuleux commented on milatools/distributed/__init__.py:
```diff
@@ -0,0 +1,181 @@
+
+    @property
+    def device(self):
+        """Return the device this scrupt should use"""
```
```suggestion
        """Return the device this script should use"""
```

breuleux reviewed (commented):
This could use some usage example to highlight the benefits.

Also, it should probably live under `milatools.torch.distributed` so that we have a fair organization of code for all frameworks (in case we want to add recipes for them).

Delaunay commented:
> This could use some usage example to highlight the benefits.
> 
> Also, it should probably live under `milatools.torch.distributed` so that we have a fair organization of code for all frameworks (in case we want to add recipes for them).

Moved to torch folder

> This could use some usage example to highlight the benefits.

I added an example

Delaunay commented:
We can make it even easier if we move some of the dataset downloading logic inside our dataset wrappers
